--- django-svn/django/contrib/auth/forms.py	2009-01-16 18:36:49.000000000 +0000
+++ django-1.0.2/django/contrib/auth/forms.py	2009-01-16 18:35:40.000000000 +0000
@@ -7,11 +7,13 @@
 from django.utils.translation import ugettext_lazy as _
 from django.utils.http import int_to_base36
 
+username_regex = r'^[\w@.]+$'
+
 class UserCreationForm(forms.ModelForm):
     """
     A form that creates a user, with no privileges, from the given username and password.
     """
-    username = forms.RegexField(label=_("Username"), max_length=30, regex=r'^\w+$',
+    username = forms.RegexField(label=_("Username"), max_length=30, regex=username_regex,
         help_text = _("Required. 30 characters or fewer. Alphanumeric characters only (letters, digits and underscores)."),
         error_message = _("This value must contain only letters, numbers and underscores."))
     password1 = forms.CharField(label=_("Password"), widget=forms.PasswordInput)
@@ -44,7 +46,7 @@
         return user
 
 class UserChangeForm(forms.ModelForm):
-    username = forms.RegexField(label=_("Username"), max_length=30, regex=r'^\w+$',
+    username = forms.RegexField(label=_("Username"), max_length=30, regex=username_regex,
         help_text = _("Required. 30 characters or fewer. Alphanumeric characters only (letters, digits and underscores)."),
         error_message = _("This value must contain only letters, numbers and underscores."))
     
